File: Java/Questions/Pizza.java
import java.util.List;
import java.util.Map;

// Create an enum to define the different sizes of pizza
enum Size {
    // Define the basic pizza sizes
    SMALL("Small"),
    MEDIUM("Medium"),
    LARGE("Large");

    private final String value;

    Size(String value) {
        this.value = value;
    }

    public String getValue() {
        return value;
    }
}

// Create the base pizza class for additional attributes
public class Pizza {
    protected String type;
    protected Size size;
    protected List<String> toppings;

    protected Map<String, Double> sizePrices = Map.of(
        "Small", 10.0,
        "Medium", 15.0,
        "Large", 20.0
    );

    protected double toppingPrice = 1.5;

    public Pizza(String type, Size size, List<String> toppings) {
        this.type = type;
        this.size = size;
        this.toppings = toppings;
    }

    public double calculatePrice() {
        if (!sizePrices.containsKey(size.getValue())) {
            throw new IllegalArgumentException("Invalid size: " + size.getValue());
        }
        double basePrice = sizePrices.get(size.getValue());
        double toppingsTotal = toppingPrice * toppings.size();
        return basePrice + toppingsTotal;
    }

    public String getType() {
        return type;
    }

    public Size getSize() {
        return size;
    }
}

// Define interfaces for Pizzas 
interface Bakeable {
    void prepare();
}

interface Grillable {
    void prepare();
}

// Create a bakeable Pizza - specifically Margerita
class MargheritaPizza extends Pizza implements Bakeable {
    // Initialize the pizza with size and list of toppings
    public MargheritaPizza(Size size, List<String> toppings) {
        super("margherita", size, toppings);
        sizePrices = Map.of(
            "Small", 12.0,
            "Medium", 16.0,
            "Large", 22.0
        );
        toppingPrice = 2.0;
    }

    // Define the prepare() method since we called the Bakeable interface
    @Override
    public void prepare() {
        System.out.println("Baking Margherita pizza...");
    }
}

// Create a grillable Pizza - specifically Pepperoni
class PepperoniPizza extends Pizza implements Grillable {
    // Initialize the pizza with size and list of toppings
    public PepperoniPizza(Size size, List<String> toppings) {
        super("pepperoni", size, toppings);
        sizePrices = Map.of(
            "Small", 14.0,
            "Medium", 18.0,
            "Large", 25.0
        );
        toppingPrice = 2.0;
    }

    // Define the prepare() method since we called the Grillable interface
    @Override
    public void prepare() {
        System.out.println("Grilling pepperoni pizza...");
    }
}

// Create the payment method interface which will process payments
interface PaymentMethod {
    void processPayment();
    String getType();
}

// Create a concrete credit card class which is a type of payment method
class CreditCard implements PaymentMethod {
    @Override
    public void processPayment() {
        System.out.println("Processing payment with Credit Card...");
    }

    @Override
    public String getType() {
        return "credit";
    }
}

// Create a concrete paypal class which is a type of payment method
class PayPal implements PaymentMethod {
    @Override
    public void processPayment() {
        System.out.println("Processing payment with PayPal...");
    }

    @Override
    public String getType() {
        return "paypal";
    }
}

// Create a concrete cash class which is a type of payment method
class Cash implements PaymentMethod {
    @Override
    public void processPayment() {
        System.out.println("Processing cash payment...");
    }

    @Override
    public String getType() {
        return "cash";
    }
}

File: Python/Questions/Pizza.py
from abc import ABC, abstractmethod
import enum
from typing import List


# Creating an enum to store the only way to define sizes
class Size(enum.Enum):
    SMALL = "Small"
    MEDIUM = "Medium"
    LARGE = "Large"
    
class Pizza:
    def __init__(self, type: str, size: Size, toppings: List[str]):
        # Core pizza attributes
        self.type = type
        self.size = size
        self.toppings = toppings
        
        # Default pricing details
        self.size_prices = {"Small": 10, "Medium": 15, "Large": 20}
        self.topping_price = 1.5


    def calculate_price(self) -> float:
        # Check if the size is valid using the size's value
        if self.size.value not in self.size_prices:
            raise ValueError(f"Invalid size: {self.size.value}")
        
        # Calculate the base price and topping price
        base_price = self.size_prices[self.size.value]
        toppings_total = self.topping_price * len(self.toppings)
        return base_price + toppings_total
    

# Define a baking interface which only applies to some pizza types
class Bakeable(ABC):
    @abstractmethod
    def prepare(self):
        pass

# Define a baking interface which only applies to some pizza types
class Grillable(ABC):
    @abstractmethod
    def prepare(self):
        pass

# Define a pizza subclass that also requires a prepare method for baking
class MargheritaPizza(Pizza, Bakeable):
    def __init__(self, size: Size, toppings: List[str]):
        super().__init__(type="margherita", size=size, toppings=toppings)
        self.size_prices = {"Small": 12, "Medium": 16, "Large": 22}
        self.topping_price = 2
        
    def prepare(self):
        print("Baking Margherita pizza...")
        
# Define a pizza subclass that also requires a prepare method for baking
class PepperoniPizza(Pizza, Grillable):
    def __init__(self, size: Size, toppings: List[str]):
        super().__init__(type="pepperoni", size=size, toppings=toppings)
        self.size_prices = {"Small": 14, "Medium": 18, "Large": 25}
        self.topping_price = 2
        
    def prepare(self):
        print("Grilling pepperoni pizza...")
        
# Define a pizza subclass that also requires a prepare method for baking
class GlutenFreePizza(Pizza):
    def __init__(self, size: Size, toppings: List[str]):
        super().__init__(type="gluten-free", size=size, toppings=toppings)
        self.size_prices = {"Small": 14, "Medium": 18, "Large": 25}
        self.topping_price = 0
    
# Create an interface that creates many different payment types
class PaymentMethod(ABC):
    def __init__(self, type: str):
        self.type = type
        
    @abstractmethod
    def process_payment(self):
        pass
        
# Define a credit card payment type
class CreditCard(PaymentMethod):
    def __init__(self):
        super().__init__(type="credit")
    
    def process_payment(self):
        print("Processing payment with Credit Card...")

# Define a paypal payment type
class PayPal(PaymentMethod):
    def __init__(self):
        super().__init__(type="paypal")
    
    def process_payment(self):
        print("Processing payment with PayPal...")
        
# Define a cash payment type
class Cash(PaymentMethod):
    def __init__(self):
        super().__init__(type="cash")
    
    def process_payment(self):
        print("Processing cash payment...")
        
# Create an orde class that only accepts a payment method
class Order:
    def __init__(self, items: List[Pizza], payment: PaymentMethod):
        self.items = items
        self.payment = payment
        
    def add_pizza(self, pizza: Pizza) -> None:
        self.items.append(pizza)
        
    def remove_pizza(self, pizza: Pizza) -> None:
        self.items.remove(pizza)
        
    def calculate_total(self) -> float:
        total = 0
        for pizza in self.items:
            total += pizza.calculate_price()
        return total
    
    def display_order_summary(self) -> None:
        print("Order Summary:")
        for pizza in self.items:
            print(f"- {pizza.type} ({pizza.size}): ${pizza.calculate_price():.2f}")
        print(f"Total: ${self.calculate_total():.2f}")
    